src/segment-loader.js

```diff
@@ -718,78 +719,65 @@ export default class SegmentLoader extends videojs.EventTarget {
-
-            if (newEstimatedDelay < rebufferingDelay) {
-              // we can switch to the playlist and avoid any rebuffering!
-              break;
-            }
-          }
-
-          // Finally, only abort early and trigger the downswitch if the time to switch
-          // is less than the time to finish the current request.
-          if (newEstimatedDelay < estimatedDelay) {
-            // set the bandwidth to that of the desired playlist adding one so the
-            // playlist selector does not exclude it)
-            this.bandwidth = safeBandwidth;
-            this.abort();
-            this.trigger('bandwidthupdate');
-            return true;
-          }
-        }
-      }
+    // Do not consider aborting early if the player is paused
+    if (this.hls_.tech_.paused() ||
+        // Or if the current playlist is on the lowestEnabledRendition
+        // TODO: Replace timeout with a boolean indicating whether this playlist is the
+        //       lowestEnabledRendition.
+        !this.xhrOptions_.timeout ||
+        // Or if the the current playlist does not have bandwidth information that can
+        // be used in the estimation calculations below
+        !(this.playlist_.attributes && this.playlist_.attributes.BANDWIDTH)) {
+      return false;
     }
+
+    const firstByteReceived = Date.now() - stats.firstByteReceived;
+
+    // Wait at least 1 second before using the calculated bandwidth from the
+    // progress event to allow the bitrate to stabilize
+    if (firstByteReceived < 1000) {
+      return false;
+    }
+
+    const measuredBandwidth = stats.bandwidth;
+    const estimatedSegmentSize =
+      this.pendingSegment_.duration * this.playlist_.attributes.BANDWIDTH;
+    const currentTime = this.currentTime_();
+    const estimatedRoundTripTime =
+      (estimatedSegmentSize - (stats.bytesReceived * 8)) / measuredBandwidth;
+    const buffered = this.buffered_();
+    const bufferedEnd = buffered.length ? buffered.end(buffered.length - 1) : 0;
+    const timeUntilRebuffer =
+      (bufferedEnd - currentTime) / this.hls_.tech_.playbackRate();
+
+    // Only consider aborting early if the estimated time to finish the download
+    // is larger than the estimated time until the player runs out of forward buffer
+    if (estimatedRoundTripTime < timeUntilRebuffer) {
+      return false;
+    }
+
+    const { playlist, roundTripTime } =
+      minRebufferingSelector(this.hls_.playlists.master,
+                             measuredBandwidth,
+                             timeUntilRebuffer,
+                             this.pendingSegment_.duration);
+
+    // Finally, only abort early and trigger the downswitch if the playlist
+    // that can minimize the time spent rebuffering is a different playlist than the
+    // current, and that switching will be less time than just waiting for the
+    // current request to complete.
+    if (playlist &&
```

Comment:
Minor, but since the other conditionals do the `return false` early, otherwise continue, may look nicer to check for cases we want to return with false instead, then the final return is true.